File: clojure/fibonacci.clj
(defn fibonacci [n]
  (let [makeF (fn [f]
                (fn [x]
                  (cond
                   (= x 0) 0
                   (= x 1) 1
                   :else   (+ (f (- x 1)) (f (- x 2))))))]
    (let [y (fn [y] (fn [x] ((makeF (y y)) x)))]
      ((y y) n))))

File: dotnet/fsharp/fibonacci.fs
module fibonacci

type Y = { call: Y -> (int -> int) }

let fibonacci n =
    let makeF f: int -> int =
        fun x ->
            if x = 0 then 0 else if x = 1 then 1 else f(x - 1) + f(x - 2)
    let y = { call = fun y -> fun x -> makeF(y.call(y)) x }
    (y.call y) n
